### DISCUSSION TITLE
Privacy when using SHAP on health data

### DISCUSSION THREAD

**Discussion by martinwellman** [shap/shap#3245] [Q&A]
I'm looking for advice on using SHAP on PHI/health data. An issue we're having is that using a SHAP scatter plot (dependence plot) it's possible to retrieve individual values for the variable on the x axis (eg. an individual's age). Our problem is that this is considered individual level data (rather than aggregate). For our purposes, taking a dot on the scatter plot and saying this individual is 48 years old (along with a lot of other 48 year old individuals) is different than saying there are 100 or so people aged 48, or 100 people whose average age is 48. Our policies state that reporting individual level data is not allowed. Our current plan is to cluster points on the scatter plot into individual aggregate points that represent a group of individuals, we'll likely just take the mean/mode/median/etc of the x-axis variable as well as the coloring variable, we'll have to decide the best way to group the points on the SHAP plot into these aggregate points. Has anyone else had to deal with this and what solution did you have for it?

**Comment by znacer** [comment]
Hi,
You could over write the explanation object to remove your data, that way you won't have colors in your plots.
'shap_values.data = None'

**Reply by znacer** [reply]
An other option you could consider is to use violin plots instead of beeswarm. This way you won't have dots representing individuals.

**Reply by martinwellman** [reply]
Thanks @znacer, unfortunately the problem isn't related to the coloring. The problem is being able to extract even just one variable for a single dot, which is possible even without the interaction term in a partial dependence plot (ie. you just need to read the x-axis in a scatter plot). I'll update the question to make it clearer. Something similar to a violin plot might be useful for a dependence plot (plotted with shap.plots.scatter), but I don't think that's been implemented. It's something we might consider trying to implement.

**Comment by znacer** [comment]
Plot a hull to a scatter plot is not an easy task. I tried to do that by splitting the data to plot in bins and then only keep the min and max. This way there is 2 line plot :
```py
#%%
import shap
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from xgboost import XGBRegressor
#%%
# load diabetes data
X,y = shap.datasets.diabetes()
model = XGBRegressor().fit(X, y)

# compute SHAP values
explainer = shap.Explainer(model)
shap_values = explainer(X)
#%%
# scatter plot
feature = "bmi"
feature_id = X.columns.tolist().index(feature) 
plot_df = pd.DataFrame({
    "features": shap_values[:, feature_id].data,
    "shap_values": shap_values[:, feature_id].values
})
plot_df.sort_values(by="features", inplace=True)

plot_df_mins = plot_df.groupby(pd.cut(plot_df["features"], bins=10)).min()
plot_df_maxs = plot_df.groupby(pd.cut(plot_df["features"], bins=10)).max()

fig, ax = plt.subplots()
plt.plot(plot_df_mins["features"], plot_df_mins["shap_values"], color="red")
plt.plot(plot_df_maxs["features"], plot_df_maxs["shap_values"], color="blue")

sns.scatterplot(data=plot_df, x="features", y="shap_values", ax=ax)
plt.show()
```
![image](https://github.com/shap/shap/assets/45372115/7fb2482e-2b2d-4245-a56b-c8d736000a85)
